libcxx/include/string

```diff
@@ -1990,10 +1998,12 @@ private:
   }
 
   _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 pointer __get_long_pointer() _NOEXCEPT {
+    _LIBCPP_ASSERT_INTERNAL(__rep_.__l.__is_long_, "String has to be long when trying to get the long pointer");
     return _LIBCPP_ASAN_VOLATILE_WRAPPER(__rep_.__l.__data_);
   }
 
   _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 const_pointer __get_long_pointer() const _NOEXCEPT {
+    _LIBCPP_ASSERT_INTERNAL(__rep_.__l.__is_long_, "String has to be long when trying to get the long pointer");
```

Comment:
Heads up!

This line (just this like) breaks out tests in  strange way. Code takes impossible branch.

I suspect either UB in this code, or pre-existing miscompile in clang.

Note: 
`_LIBCPP_ASSERT_INTERNAL(__is_long(), ...`
fixes the issue.

Does this ring a bell?

Reply:
You mean that after landing this patch, some internal tests unrelated to `string` (in particular, not libc++ tests) start breaking?

That's really strange, indeed there may be some subtle UB that the compiler is taking advantage of here.

Reply:
One thing I could imagine is that if somehow this assertion is not satisfied and the compiler can tell at compile-time, then this check itself might be UB because accessing `__rep_.__l` is invalid (that member of the union is not active). @philnik777 thoughts?

Reply:
We're doing basically exactly the same in `__is_long()`, so I'm not sure how that would change the compiler's behaviour. FWIW Clang and GCC allow accessing an inactive union member in this kind of a scenario, so I don't think that's the problem.

Reply:
@vitalybuka Do you think you'd be able to get us a reproducer? I'd really like to be able to investigate that, but it's the kind of thing that's going to be difficult to make progress on without a repro.

Reply:
I tried to bisect llvm revisions and it still reproducible >1 years back, if I insert just a single equivalent of 
`_LIBCPP_ASSERT_INTERNAL(__rep_.__l.__is_long_, "String has to be long when trying to get the long pointer");`

Something like

```
# just to normalize for sed
clang-format -i llvm/include/c++/v1/string

sed -E -i 's/(__get_long_pointer\(\) .*_NOEXCEPT \{)/\1 _LIBCPP_ASSERT_INTERNAL(__r_.first().__l.__is_long_, "");/g' llvm/include/c++/v1/string
```

I am in progress of creduce example, but I rather suspect a compiler bug.

Reply:
Sorry, false alarm.
Out code had UB, which now behaves differently.
Variadic function was called with wrong number of arguments, so it's unpacked wrong value, which was affected by std::string changes.

Reply:
Thanks for the update!